0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+As the key is only 31 bytes, i.e. 248 bits, there are still 5 bits of data that
+can be used to encode an index of the data for a specific key. Therefore, for a
+given pre-image key, we can encode 2^5 * 31 bytes, i.e. 992 bytes. In terms of
+constraints, it will cost an additional constraint to mask the key and get the
+corresponding index. Using this encoding trick, and supposing the table can be
+of maximum size 2^16, we can encode 2048 different pre-image keys in one table.
+
+In addition to this technique, a wider table can also be used to encode bigger data.
+
+<!-- TODO: a word on the pre-image key as we can only read 4 bytes by 4 bytes,
+     we need to save it somewhere -->
+
+### `sysMmap`
+
+<!-- DONE -->
+
+This can be ignored at the moment and implemented as a no-op.
+For consistency with Cannon, `v0` must be set to the value of the register `a0`
+and `a3` must be set to `0`.
+
+### `sysBrk`
+
+<!-- DONE -->
+
+The register `v0` must be set to `0x400000` and the register `a3` to `0`.
+
+### `sysClone`
+
+<!-- DONE -->
+
+The register `v0` must be set to `1` and the register `a3` to `0`.
+
+### `sysExitGroup`
+
+<!-- DONE -->
+
+The program must halt in this case and `set_halted` must be called with the value `a0`.
+
+### `sysRead`
+
+This syscall is used to get data from the outside world and write in the VM memory.
+
+The register `a0` contains the file descriptor of the input channel, the
+register `a1` the address and the `a2` contains the number of bytes that must be
+read.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read, and `a3` must contain the error code.
+
+Different input channels must be supported.
+
+#### Standard input
+
+<!-- DONE -->
+
+In the case of the standard input, i.e. when `a0` equals `0`, nothing in
+particular happens. Therefore, the registers `v0` and `a3` are set to `0`, which
+corresponds to no data has been read and no error is raised.
+
+#### Pre-image oracle
+
+When the buffer to be read is the pre-image oracle, i.e. when `a0` equals `5`, a
+chunk of the pre-image is read, and copy into the memory.
+The pre-image that must be read is determined by the `preimageKey`, and the actual
+chunk is given by the preimage offset variable.
+
+The number of bytes that must be read is in the register `a2` and the address
+where the chunk as to be written is in the register `a1`.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read. Also, the preimage offset has to be incremented by the number of
+bytes that has been read.
+
+The preimage data length is located in the 8 first addresses, encoded on 64
+bits in big-endian.
+
+A difference with Cannon is we verify the preimage data is the actual data the
+user has requested previously. To achieve this, we will use the optimised
+keccak gate described in [Keccak RFC](./keccak.md).
+
+The keccak sponge state is 136 bytes, and is located in the columns `[100,
```

Comment:
Actually it is 200 bytes (1600 bits). It's just that at the beginning of the absorb, the last 64 bytes are zeros (meaning, they don't correspond to preimage data).